### DISCUSSION TITLE
TestOpen test cases

### DISCUSSION THREAD

**Discussion by aerinng** [ANLAB-KAIST/KENSv3#57] [Project Q&A 2. Transport Layer (TCP)]
This is for case SOCKET. In the skeleton code, I noticed there is this line `SystemCallInterface(AF_INET, IPPROTO_TCP, host)`. PA2 also mentioned that "In KENS, you need to implement only domain AF_INET, type SOCK_STREAM, and protocol IPPROTO_TCP." Hence I want to check if we are required to pass in the domain, type and protocol to implement case SOCKET as I don't seem to see `type` in `SystemCallInterface`? Or is `createFileDescriptor(pid)` sufficient as I cannot seem to pass all the test cases. Thank you!

**Comment by Pusnow** [comment]
https://github.com/ANLAB-KAIST/KENSv3/blob/0226d32cc49a1a2df8c2a3d613ba8e7509502549/src/Networking/TCP/E_TCPApplication.cpp#L14-L21


KENS ignores the `type` option and never uses it. 



https://github.com/ANLAB-KAIST/KENSv3/blob/0226d32cc49a1a2df8c2a3d613ba8e7509502549/src/Networking/E_Host.cpp#L117


KENS only uses `domain` and `protocol` to figure out which networking stack (TCP or UDP) to use.





> I want to check if we are required to pass in the domain, type and protocol to implement case SOCKET as I don't seem to see type in SystemCallInterface? Or is createFileDescriptor(pid) sufficient as I cannot seem to pass all the test cases. 

No, you don't need to do so. You can assume all `socket` system callback in the `TCPAssignment` is `AF_INET `, `SOCK_STREAM `, and `IPPROTO_TCP `. Also,`createFileDescriptor(pid)` is sufficient.

**Reply by aerinng** [reply]
I only used `createFileDescriptor(pid)` in my returnSystemCall but I failed the test cases. May I ask for some help?

<img width="642" alt="image" src="https://user-images.githubusercontent.com/49832045/161676383-208fdb08-3c4e-425e-8aa0-00f214b70f40.png">

**Reply by Pusnow** [reply]
See #56. You should implement `close()` also.

**Reply by aerinng** [reply]
I have implemented `close()` and timer and still get the same error. I would like to check if we are required to catch any errors from `close()` and  then do `returnSystemCall(syscallUUID, -1)`? Or is returning `returnSystemCall(syscallUUID, 0)` sufficient for `close()`?

**Reply by Pusnow** [reply]
Your fail count is `130048`. In the test case, all `socket` calls must succeed.

>  I would like to check if we are required to catch any errors from close() and then do returnSystemCall(syscallUUID, -1)?

Which error? There should be no errors.


The test case should proceed like this:

1. call `socket()` 1024 times. They return positive file descriptors (checked by next if statement). 
2. check duplicated file descriptors. There should no duplication.
3. close all file descriptors.
4. Goto 1 and repeat 128 times. Since all file descriptors are closed. Same numbers will be returned in the first step.
5. Lastly, check the number of success (must be 1024*128), fail (must be 0), and duplication (must be 0).
